src/day8/re_frame/test.cljc

```diff
@@ -1,66 +1,242 @@
-             (cb new-event))))))))
-
-(defn wait-for-ch
-  [ids]
-  (let [id-set (normalize-ids ids)
-        ch (async/chan)
-        cb-id (gensym "wait-for-ch-cb-fn")]
-    (r/add-post-event-callback
-      cb-id
-      (fn listener [new-event _]
-        (let [new-id (first new-event)]
-          (when (get id-set new-id)
-            (async/put! ch new-event)
-            (async/close! ch)
-            (r/remove-post-event-callback cb-id)))))
-    ch))
+  #?(:cljs (:require-macros day8.re-frame.test))
+  (:require #?(:cljs [cljs.test :as test]
+               :clj  [clojure.test :as test])
+            [re-frame.core :as rf]
+            [re-frame.db :as rf-db]
+            [re-frame.interop :as rf-int]
+            [re-frame.registrar :as rf-registrar])
+  #?(:clj (:import [java.util.concurrent Executors TimeUnit])))
+
+
+;;;;
+;;;; General test utils
+;;;;
+
+(defn- dequeue!
+  "Dequeue an item from a persistent queue which is stored as the value of the
+  given atom, setting the atom to the new queue (minus dequeued item) and
+  returning the dequeued item. If the queue is empty, does not alter it and
+  returns nil."
+  [queue-atom]
+  (let [queue @queue-atom]
+    (when (seq queue)
+      (if (compare-and-set! queue-atom queue (pop queue))
+        (peek queue)
+        (recur queue-atom)))))
+
+
+(defn make-re-frame-restore-fn []
+  (let [handlers @@#'rf-registrar/kind->id->handler
+        app-db   @@#'rf-db/app-db]
+    (fn []
+      (reset! @#'rf-registrar/kind->id->handler handlers)
+      (reset! @#'rf-db/app-db app-db)
+      nil)))
+
+(defmacro with-restored-re-frame-state
+  "Run `body`, but discard whatever effects it may have on re-frame's internal
+  state (by resetting `app-db` and re-frame's various different types of
+  handlers after `body` has run)."
+  [& body]
+  `(let [restore-fn# (make-re-frame-restore-fn)]
+     (try
+       ~@body
+       (finally
+         (restore-fn#)))))
+
+
+;;;;
+;;;; Async tests
+;;;;
+
+(def ^:dynamic *test-timeout* 5000)
+
+(def ^{:dynamic true, :private true} *test-context*
+  "`*test-context*` is used to communicate internal details of the test between
+  `run-test-async*` and `wait-for*`. It is dynamically bound so that it doesn't
+  need to appear as a lexical argument to a `wait-for` block, since we don't
+  want it to be visible when you're writing tests.  But care must be taken to
+  pass it around lexically across callbacks, since ClojureScript doesn't have
+  `bound-fn`."
+  nil)
+
+
+(defn run-test-async* [f]
+  (let [test-context {:wait-for-depth     0
+                      :max-wait-for-depth (atom 0)
+                      :now-waiting-for    (atom nil)}]
+    #?(:clj  (let [done-promise (promise)
+                   executor     (Executors/newSingleThreadExecutor)
+                   fail-ex      (atom nil)
+                   test-context (assoc test-context :done #(deliver done-promise ::done))]
+               (with-redefs [rf-int/executor executor]
```

Comment:
we need to be careful about with-redefs combined with core-async. Don't think it's an issue here though.

https://nvbn.github.io/2014/11/05/protocols-for-testing/
http://dev.clojure.org/jira/browse/CLJS-884